README.md
# dnscrypt-proxy2 for Android [Revival] 
### [bluemeda's original magisk module - Revived](https://github.com/Magisk-Modules-Repo/dnscrypt-proxy/) 

A flexible DNS proxy, with support for modern encrypted DNS protocols such as [DNSCrypt v2](https://dnscrypt.info/protocol), [DNS-over-HTTPS](https://www.rfc-editor.org/rfc/rfc8484.txt) and [Anonymized DNSCrypt](https://github.com/DNSCrypt/dnscrypt-protocol/blob/master/ANONYMIZED-DNSCRYPT.txt).

## Features
- arm, arm64, x86 and x86_64 are supported.
- ipv4  and ipv6  are supported.
- All binary files are downloaded from [https://github.com/jedisct1/dnscrypt-proxy/releases](https://github.com/jedisct1/dnscrypt-proxy/releases)

## Installation
- Download ~~[.zip module](https://github.com/Magisk-Modules-Repo/dnscrypt-proxy2/releases)~~ / flash it in Magisk Manager App  and set the DNS server as follows. 

### Set DNS server manually with 3rd-party app (not included in this module)
- DNS server address is 127.0.0.1:5354 for ipv4 and [::1]:5354 for ipv6
- If you use AfWall, you can write this custom packet  redirection script where IP address 9.9.9.9 (can be 1.1.1.1 or 8.8.8.8 or any IP address of a reliable DNS server) should be same as the IP address of fallback_resolver & netprobe_address in configuration file dnscrypt-proxy.toml located at /data/media/0/dnscrypt-proxy. 
  ```
  iptables -t nat -A OUTPUT -p tcp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination 127.0.0.1:5354
  iptables -t nat -A OUTPUT -p udp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination 127.0.0.1:5354
  ``` 

~~ip6tables -t nat -A OUTPUT -p tcp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination [::1]:5354~~


~~ip6tables -t nat -A OUTPUT -p udp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination [::1]:5354~~


and this shutdown script
  ```
  iptables -t nat -D OUTPUT -p tcp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination 127.0.0.1:5354
  iptables -t nat -D OUTPUT -p udp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination 127.0.0.1:5354
  ```
 
~~ip6tables -t nat -D OUTPUT -p tcp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination [::1]:5354~~


~~ip6tables -t nat -D OUTPUT -p udp ! -d 9.9.9.9 --dport 53 -j DNAT --to-destination [::1]:5354~~


Refer AFWall [Docs](https://github.com/ukanth/afwall/wiki), [FAQs](https://github.com/ukanth/afwall/wiki/FAQ) and [custom scripts](https://github.com/ukanth/afwall/wiki/CustomScripts). 

A script file named dns-redirect.sh is included in the configuration directory. The file may be copied to /data/adb/service.d & its permission set to executable for auto-redirection of dns request (for those users not interested in using  AFWall or other apps).

  
DNS66 or dnsfilter app (both apps available in F-Droid repo) can also  be used instead of custom script in  AFWall. 

## Configuration (post-installation)
- Read & correct the configuration options, especially the IP address, which should be same in redirection script and fallback_resolver & netprobe_address of dnscrypt-proxy.toml. 
- Configuration is located at `/sdcard/dnscrypt-proxy/dnscrypt-proxy.toml` [or /data/media/0/dnscrypt-proxy/dnscrypt-proxy.toml]
- Changes in configuration takes effect only after the dnscrypt-proxy [or the phone] is restarted. Typing "pkill -HUP dnscrypt-proxy" in terminal restarts the dnscrypt-proxy process [without rebooting the phone].
- For more detailed configuration please refer to [upstream documentation](https://github.com/jedisct1/dnscrypt-proxy/wiki/Configuration)

## Testing
- [http://dnssec.vs.uni-due.de/](http://dnssec.vs.uni-due.de/) 
- [http://www.dnssec-or-not.com/](http://www.dnssec-or-not.com/) online test by VeriSign (no JavaScript required)
- [http://en.conn.internet.nl/connection/](http://en.conn.internet.nl/connection/) online test by Dutch Internet Standards Platform
- [http://www.dnssec-failed.org/](http://www.dnssec-failed.org/) webpage with bogus signature by Comcast (will not open at all if you are using DNSSEC)
- [http://dnsleak.com/](http://dnsleak.com/)
- [https://www.dnsleaktest.com/](https://www.dnsleaktest.com/)
## Changelog
- [Full changelog](https://github.com/Magisk-Modules-Repo/dnscrypt-proxy2/blob/master/changelog.md)
## Credits
- DNSCrypt-Proxy2 upstream | [jedisct1](https://github.com/jedisct1/dnscrypt-proxy)
- [bluemeda & all other contributors for the magisk module](https://github.com/Magisk-Modules-Repo/dnscrypt-proxy2/graphs/contributors)


changelog.md
# Changelog
## Changelog
### v2.9.3
- updated magisk module template
- updated binary & configuration files to upstream version 2.0.36

### v2.9.2
- updated binary & configuration files to upstream version 2.0.33
### v2.9.1
- Revival of dnscrypt-proxy2  magisk-module of bluemeda 
- updated binary & configuration files to upstream version 2.0.31
- redirection script included for manual copy to bootup auto-start directory. 
- updated README.MD
### v2.9.0
- Revival of dnscrypt-proxy2  magisk-module of bluemeda 
- updated binary & configuration files to upstream version 2.0.29
### v2.8.9
- Revival of dnscrypt-proxy2  magisk-module of bluemeda 
- updated binary & configuration files to upstream version 2.0.28
### v2.8.7
- changed path of configuration file [dnscrypt-toml] from /system/etc/ to /data/media/0/ [or /sdcard]
- updated binary & configuration files to 2.0.25
- Removed automatic redirection of dns-request and let dnscrypt-proxy do its job only.
### v2.8.5
- Fix #40
### v2.8.4
- Fix failed to copy or backup config file
### v2.8.3
- Fix permission issue
- Add option to replace or backup-restore config file
### v2.8.2
- Fix "binary file is missing"
### v2.8.1
- Update Magisk 18100 requirements
### v2.8.0
- Update binary files 2.0.22
### v2.7.0
- Update binary files 2.0.21
### v2.6.0
- Update binary files to 2.0.19
### v2.5.0
- Update binary files to 2.0.16
- add exception for cloudflare fallback resolver.
## v2.4.0
* Update binary files to 2.0.14
## v2.3.0
* Update binary files to 2.0.10 
* Add option to choose auto redirect DNS or manually set with 3rd-party app.
## v2.2.0
* Update binary files to 2.0.8
## v2.1.3
* __If you have previous version, please uninstall it first then reinstall it again or you can change listen port manually in `dnscrypt-proxy.toml` file.__
* Fix Tethering Client cannot Resolve DNSCrypt
* Fix Chromecast devices not showing [jedisct1/dnscrypt-proxy#226](https://github.com/jedisct1/dnscrypt-proxy/issues/226)
* Add binary files for x86 and x86_64 (test)
## v2.1.2
* Bug Fixes
## v2.1.1
* Bug fixes
## v2.1
* Bug fixes
## v2.0
* Resolve download.dnscrypt.info first before executing `iptables`
* Don't override dnscrypt-proxy.toml if exist
* Update binary files to v2.0.6
## v1.1
* Change listen port to 5353 (avoid conflict while tethering)
## v1.0
* Initial release
* dnscrypt-proxy v2.0.5



install.sh
##########################################################################################
#
# Magisk Module Installer Script
#
##########################################################################################
##########################################################################################
#
# Instructions:
#
# 1. Place your files into system folder (delete the placeholder file)
# 2. Fill in your module's info into module.prop
# 3. Configure and implement callbacks in this file
# 4. If you need boot scripts, add them into common/post-fs-data.sh or common/service.sh
# 5. Add your additional or modified system properties into common/system.prop
#
##########################################################################################

##########################################################################################
# Config Flags
##########################################################################################

# Set to true if you do *NOT* want Magisk to mount
# any files for you. Most modules would NOT want
# to set this flag to true
SKIPMOUNT=false

# Set to true if you need to load system.prop
PROPFILE=false

# Set to true if you need post-fs-data script
POSTFSDATA=false

# Set to true if you need late_start service script
LATESTARTSERVICE=true

##########################################################################################
# Replace list
##########################################################################################

# List all directories you want to directly replace in the system
# Check the documentations for more info why you would need this

# Construct your list in the following format
# This is an example
REPLACE_EXAMPLE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here
REPLACE="
"

##########################################################################################
#
# Function Callbacks
#
# The following functions will be called by the installation framework.
# You do not have the ability to modify update-binary, the only way you can customize
# installation is through implementing these functions.
#
# When running your callbacks, the installation framework will make sure the Magisk
# internal busybox path is *PREPENDED* to PATH, so all common commands shall exist.
# Also, it will make sure /data, /system, and /vendor is properly mounted.
#
##########################################################################################
##########################################################################################
#
# The installation framework will export some variables and functions.
# You should use these variables and functions for installation.
#
# ! DO NOT use any Magisk internal paths as those are NOT public API.
# ! DO NOT use other functions in util_functions.sh as they are NOT public API.
# ! Non public APIs are not guranteed to maintain compatibility between releases.
#
# Available variables:
#
# MAGISK_VER (string): the version string of current installed Magisk
# MAGISK_VER_CODE (int): the version code of current installed Magisk
# BOOTMODE (bool): true if the module is currently installing in Magisk Manager
# MODPATH (path): the path where your module files should be installed
# TMPDIR (path): a place where you can temporarily store files
# ZIPFILE (path): your module's installation zip
# ARCH (string): the architecture of the device. Value is either arm, arm64, x86, or x64
# IS64BIT (bool): true if $ARCH is either arm64 or x64
# API (int): the API level (Android version) of the device
#
# Availible functions:
#
# ui_print <msg>
#     print <msg> to console
#     Avoid using 'echo' as it will not display in custom recovery's console
#
# abort <msg>
#     print error message <msg> to console and terminate installation
#     Avoid using 'exit' as it will skip the termination cleanup steps
#
# set_perm <target> <owner> <group> <permission> [context]
#     if [context] is empty, it will default to "u:object_r:system_file:s0"
#     this function is a shorthand for the following commands
#       chown owner.group target
#       chmod permission target
#       chcon context target
#
# set_perm_recursive <directory> <owner> <group> <dirpermission> <filepermission> [context]
#     if [context] is empty, it will default to "u:object_r:system_file:s0"
#     for all files in <directory>, it will call:
#       set_perm file owner group filepermission context
#     for all directories in <directory> (including itself), it will call:
#       set_perm dir owner group dirpermission context
#
##########################################################################################
##########################################################################################
# If you need boot scripts, DO NOT use general boot scripts (post-fs-data.d/service.d)
# ONLY use module scripts as it respects the module status (remove/disable) and is
# guaranteed to maintain the same behavior in future Magisk releases.
# Enable boot scripts by setting the flags in the config section above.
##########################################################################################

# Set what you want to display when installing your module

print_modname() {
  ui_print " "
  ui_print "*******************************"
  ui_print "*       DNSCrypt-Proxy2       *"
  ui_print "*        Magisk Module        *"
  ui_print "*******************************"
  ui_print "*        v2.9.2               *"
  ui_print "*  created by bluemeda        *"
  ui_print "*  revived by lindroidux      *"
  ui_print "*******************************"
  ui_print " "
}

# Copy/extract your module files into $MODPATH in on_install.

on_install() {
  # The following is the default implementation: extract $ZIPFILE/system to $MODPATH
  # Extend/change the logic to whatever you want

  if [ "$ARCH" == "arm" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-arm
  elif [ "$ARCH" == "arm64" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-arm64
  elif [ "$ARCH" == "x86" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-i386
  elif [ "$ARCH" == "x64" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-x86_64
  fi

  CONFIG_PATH=$TMPDIR/config

  unzip -o "$ZIPFILE" 'config/*' 'binary/*' -d $TMPDIR

  ui_print "* Creating binary path"
  mkdir -p $MODPATH/system/bin

  ui_print "* Creating config path"
  mkdir -p /data/media/0/dnscrypt-proxy

  if [ -f "$BINARY_PATH" ]; then
    ui_print "* Copying binary for $ARCH"
    cp -af $BINARY_PATH $MODPATH/system/bin/dnscrypt-proxy
  else
    abort "Binary file for $ARCH is missing!"
  fi

  if [ -d "$CONFIG_PATH" ]; then
    ui_print "* Copying example and license files"
    cp -af $CONFIG_PATH/* /data/media/0/dnscrypt-proxy/
  else
    abort "Config file is missing!"
  fi

  . $TMPDIR/option.sh

}

# Only some special files require specific permissions
# This function will be called after on_install is done
# The default permissions should be good enough for most cases


set_permissions() {
  # The following is the default rule, DO NOT remove
  set_perm_recursive $MODPATH 0 0 0755 0644
  set_perm $MODPATH/system/bin/dnscrypt-proxy 0 0 0755
  # Here are some examples:
  # set_perm_recursive  $MODPATH/system/lib       0     0       0755      0644
  # set_perm  $MODPATH/system/bin/app_process32   0     2000    0755      u:object_r:zygote_exec:s0
  # set_perm  $MODPATH/system/bin/dex2oat         0     2000    0755      u:object_r:dex2oat_exec:s0
  # set_perm  $MODPATH/system/lib/libart.so       0     0       0644
}

# You can add more functions to assist your custom script code


customize.sh
  ui_print " "
  ui_print "*******************************"
  ui_print "*       DNSCrypt-Proxy2       *"
  ui_print "*        Magisk Module        *"
  ui_print "*******************************"
  ui_print "*        v2.9.3               *"
  ui_print "*  created by bluemeda        *"
  ui_print "*  revived by lindroidux      *"
  ui_print "*******************************"
  ui_print " "

  if [ "$ARCH" == "arm" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-arm
  elif [ "$ARCH" == "arm64" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-arm64
  elif [ "$ARCH" == "x86" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-i386
  elif [ "$ARCH" == "x64" ];then
    BINARY_PATH=$TMPDIR/binary/dnscrypt-proxy-android-x86_64
  fi

  CONFIG_PATH=$TMPDIR/config

  unzip -o "$ZIPFILE" 'config/*' 'binary/*' -d $TMPDIR

  ui_print "* Creating binary path"
  mkdir -pv $MODPATH/system/bin

  ui_print "* Creating config & docs path"
  mkdir -p /data/media/0/dnscrypt-proxy/docs-upstream

  if [ -f "$BINARY_PATH" ]; then
    ui_print "* Copying binary for $ARCH"
    cp -afv $BINARY_PATH $MODPATH/system/bin/dnscrypt-proxy
  else
    abort "Binary file for $ARCH is missing!"
  fi

  if [ -d "$CONFIG_PATH" ]; then
    ui_print "* Copying example and license files"
    cp -arfv $CONFIG_PATH/* /data/media/0/dnscrypt-proxy/
  else
    abort "Config file is missing!"
  fi

CONFIG_FILE="/data/media/0/dnscrypt-proxy/dnscrypt-proxy.toml" 

  if  [ -f "$CONFIG_FILE" ]; then
## v2.8.5
     ui_print "* Backing up config file"
    # cp -afv $CONFIG_FILE $TMPDIR
    # ui_print "* Restoring config files"
    # cp -af $TMPDIR/dnscrypt-proxy.toml $CONFIG_FILE 
     
## v2.8.5
 #   cp -afv /data/media/0/dnscrypt-proxy/example-dnscrypt-proxy.toml $CONFIG_FILE
 #  sed -i -e 's/127.0.0.1:53/127.0.0.1:5354/g' $CONFIG_FILE
#    sed -i -e 's/\[::1\]:53/\[::1\]:5354/g' $CONFIG_FILE

## v2.8.9
cp -afv  $CONFIG_FILE ${CONFIG_FILE}-backup-`date +%Y%m%d%H%M%S`
  fi

cp -afv /data/media/0/dnscrypt-proxy/my-dnscrypt-proxy.toml $CONFIG_FILE

  set_perm_recursive $MODPATH 0 0 0755 0755
  set_perm $MODPATH/system/bin/dnscrypt-proxy 0 0 0755
